File: school-frontend/src/api/libraryApi.ts
import axiosInstance from './axiosInstance';
import { ENDPOINTS } from './endpoints';
import type {
  ApiResponse,
  Book,
  BookCategory,
  BookIssue,
  BookIssueStatus,
  LibraryDashboard,
  PageResponse,
} from '@/types';

export interface BookCategoryPayload {
  name: string;
}

export interface BookListParams {
  page?: number;
  size?: number;
  search?: string;
  categoryId?: number;
  sort?: string;
}

export interface BookPayload {
  title: string;
  author: string;
  isbn: string;
  categoryId: number;
  publisher?: string;
  totalCopies: number;
  rackNumber?: string;
  price?: number;
}

export interface BookIssueListParams {
  bookId?: number;
  studentId?: number;
  teacherId?: number;
  status?: BookIssueStatus;
  page?: number;
  size?: number;
  sort?: string;
}

export interface IssueBookPayload {
  bookId: number;
  studentId?: number;
  teacherId?: number;
  dueDate: string;
}

export interface ReturnBookPayload {
  returnDate: string;
}

/** Typed wrappers around every /api/v1/book-categories/**, /books/**, /book-issues/**, /library/** endpoint. */
export const libraryApi = {
  bookCategories: {
    list: async (): Promise<ApiResponse<BookCategory[]>> => {
      const { data } = await axiosInstance.get<ApiResponse<BookCategory[]>>(ENDPOINTS.BOOK_CATEGORIES.BASE);
      return data;
    },
    create: async (payload: BookCategoryPayload): Promise<ApiResponse<BookCategory>> => {
      const { data } = await axiosInstance.post<ApiResponse<BookCategory>>(ENDPOINTS.BOOK_CATEGORIES.BASE, payload);
      return data;
    },
    update: async (id: number, payload: BookCategoryPayload): Promise<ApiResponse<BookCategory>> => {
      const { data } = await axiosInstance.put<ApiResponse<BookCategory>>(
        ENDPOINTS.BOOK_CATEGORIES.BY_ID(id),
        payload,
      );
      return data;
    },
    remove: async (id: number): Promise<ApiResponse<null>> => {
      const { data } = await axiosInstance.delete<ApiResponse<null>>(ENDPOINTS.BOOK_CATEGORIES.BY_ID(id));
      return data;
    },
  },

  books: {
    list: async (params: BookListParams = {}): Promise<ApiResponse<PageResponse<Book>>> => {
      const { data } = await axiosInstance.get<ApiResponse<PageResponse<Book>>>(ENDPOINTS.BOOKS.BASE, { params });
      return data;
    },
    getById: async (id: number): Promise<ApiResponse<Book>> => {
      const { data } = await axiosInstance.get<ApiResponse<Book>>(ENDPOINTS.BOOKS.BY_ID(id));
      return data;
    },
    create: async (payload: BookPayload): Promise<ApiResponse<Book>> => {
      const { data } = await axiosInstance.post<ApiResponse<Book>>(ENDPOINTS.BOOKS.BASE, payload);
      return data;
    },
    update: async (id: number, payload: BookPayload): Promise<ApiResponse<Book>> => {
      const { data } = await axiosInstance.put<ApiResponse<Book>>(ENDPOINTS.BOOKS.BY_ID(id), payload);
      return data;
    },
    remove: async (id: number): Promise<ApiResponse<null>> => {
      const { data } = await axiosInstance.delete<ApiResponse<null>>(ENDPOINTS.BOOKS.BY_ID(id));
      return data;
    },
  },

  bookIssues: {
    list: async (params: BookIssueListParams = {}): Promise<ApiResponse<PageResponse<BookIssue>>> => {
      const { data } = await axiosInstance.get<ApiResponse<PageResponse<BookIssue>>>(ENDPOINTS.BOOK_ISSUES.BASE, {
        params,
      });
      return data;
    },
    issue: async (payload: IssueBookPayload): Promise<ApiResponse<BookIssue>> => {
      const { data } = await axiosInstance.post<ApiResponse<BookIssue>>(ENDPOINTS.BOOK_ISSUES.ISSUE, payload);
      return data;
    },
    returnBook: async (id: number, payload: ReturnBookPayload): Promise<ApiResponse<BookIssue>> => {
      const { data } = await axiosInstance.patch<ApiResponse<BookIssue>>(ENDPOINTS.BOOK_ISSUES.RETURN(id), payload);
      return data;
    },
    overdue: async (
      params: { page?: number; size?: number; sort?: string } = {},
    ): Promise<ApiResponse<PageResponse<BookIssue>>> => {
      const { data } = await axiosInstance.get<ApiResponse<PageResponse<BookIssue>>>(ENDPOINTS.BOOK_ISSUES.OVERDUE, {
        params,
      });
      return data;
    },
  },

  dashboard: {
    get: async (): Promise<ApiResponse<LibraryDashboard>> => {
      const { data } = await axiosInstance.get<ApiResponse<LibraryDashboard>>(ENDPOINTS.LIBRARY.DASHBOARD);
      return data;
    },
  },
};

export default libraryApi;

File: school-android/app/src/main/java/com/greenwood/school/data/remote/api/LibraryApi.kt
package com.greenwood.school.data.remote.api

import com.greenwood.school.data.remote.dto.ApiEnvelope
import com.greenwood.school.data.remote.dto.BookCategoryDto
import com.greenwood.school.data.remote.dto.BookDto
import com.greenwood.school.data.remote.dto.BookIssueDto
import com.greenwood.school.data.remote.dto.BookRequestDto
import com.greenwood.school.data.remote.dto.IssueBookRequestDto
import com.greenwood.school.data.remote.dto.LibraryDashboardDto
import com.greenwood.school.data.remote.dto.NamedRequestDto
import com.greenwood.school.data.remote.dto.PageEnvelope
import com.greenwood.school.data.remote.dto.ReturnBookRequestDto
import retrofit2.http.Body
import retrofit2.http.DELETE
import retrofit2.http.GET
import retrofit2.http.PATCH
import retrofit2.http.POST
import retrofit2.http.PUT
import retrofit2.http.Path
import retrofit2.http.Query

/** `/book-categories`, `/books`, `/book-issues`, `/library/dashboard`. */
interface LibraryApi {

    @GET("book-categories")
    suspend fun getCategories(): ApiEnvelope<List<BookCategoryDto>>

    @POST("book-categories")
    suspend fun createCategory(@Body request: NamedRequestDto): ApiEnvelope<BookCategoryDto>

    @PUT("book-categories/{id}")
    suspend fun updateCategory(@Path("id") id: Long, @Body request: NamedRequestDto): ApiEnvelope<BookCategoryDto>

    @DELETE("book-categories/{id}")
    suspend fun deleteCategory(@Path("id") id: Long): ApiEnvelope<Unit>

    @GET("books")
    suspend fun getBooks(
        @Query("search") search: String? = null,
        @Query("categoryId") categoryId: Long? = null,
        @Query("page") page: Int = 0,
        @Query("size") size: Int = 20,
        @Query("sort") sort: String = "title,asc",
    ): ApiEnvelope<PageEnvelope<BookDto>>

    @GET("books/{id}")
    suspend fun getBook(@Path("id") id: Long): ApiEnvelope<BookDto>

    @POST("books")
    suspend fun createBook(@Body request: BookRequestDto): ApiEnvelope<BookDto>

    @PUT("books/{id}")
    suspend fun updateBook(@Path("id") id: Long, @Body request: BookRequestDto): ApiEnvelope<BookDto>

    @DELETE("books/{id}")
    suspend fun deleteBook(@Path("id") id: Long): ApiEnvelope<Unit>

    @GET("book-issues")
    suspend fun getIssues(
        @Query("bookId") bookId: Long? = null,
        @Query("studentId") studentId: Long? = null,
        @Query("teacherId") teacherId: Long? = null,
        @Query("status") status: String? = null,
        @Query("page") page: Int = 0,
        @Query("size") size: Int = 20,
        @Query("sort") sort: String = "issueDate,desc",
    ): ApiEnvelope<PageEnvelope<BookIssueDto>>

    @POST("book-issues/issue")
    suspend fun issueBook(@Body request: IssueBookRequestDto): ApiEnvelope<BookIssueDto>

    @PATCH("book-issues/{id}/return")
    suspend fun returnBook(@Path("id") id: Long, @Body request: ReturnBookRequestDto): ApiEnvelope<BookIssueDto>

    @GET("book-issues/overdue")
    suspend fun getOverdue(
        @Query("page") page: Int = 0,
        @Query("size") size: Int = 20,
        @Query("sort") sort: String = "dueDate,asc",
    ): ApiEnvelope<PageEnvelope<BookIssueDto>>

    @GET("library/dashboard")
    suspend fun getDashboard(): ApiEnvelope<LibraryDashboardDto>
}
